ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/ParseReportingUrl.java

```diff
@@ -87,40 +94,78 @@ public Result<PCollection<PubsubMessage>, PubsubMessage> expand(
           }
 
           Map<String, String> attributes = new HashMap<>(message.getAttributeMap());
+          String namespace = Objects
+              .requireNonNull(message.getAttribute(Attribute.DOCUMENT_NAMESPACE));
+          String docType = Objects.requireNonNull(message.getAttribute(Attribute.DOCUMENT_TYPE));
+
+          SponsoredInteraction.Builder interactionBuilder = SponsoredInteraction.builder();
+
+          interactionBuilder.setOriginalNamespace(namespace);
+          interactionBuilder.setOriginalDocType(docType);
+
+          // set fields based on namespace/doctype combos
+          if (NS_DESKTOP.equals(namespace)) {
+            interactionBuilder.setFormFactor(SponsoredInteraction.FORM_DESKTOP);
+            interactionBuilder.setInteractionType(
+                docType.contains("click") ? SponsoredInteraction.INTERACTION_CLICK
+                    : SponsoredInteraction.INTERACTION_IMPRESSION);
+          } else {
+            interactionBuilder.setFormFactor(SponsoredInteraction.FORM_PHONE);
+            ArrayNode events = payload.withArray("events");
+            if (events.size() != 1) {
+              throw new RuntimeException("expect exactly 1 event in ping.");
+            }
+            JsonNode event = events.get(0);
+            String eventName = event.path("name").asText();
+            interactionBuilder.setInteractionType(
+                eventName.contains("click") ? SponsoredInteraction.INTERACTION_CLICK
+                    : SponsoredInteraction.INTERACTION_IMPRESSION);
+          }
+
+          interactionBuilder.setSource(docType.contains(SponsoredInteraction.SOURCE_TOPSITES)
+              ? SponsoredInteraction.SOURCE_TOPSITES
+              : SponsoredInteraction.SOURCE_SUGGEST);
```

Comment:
I'm curious about this. If you don't mind expanding on what you mean in this comment.

In a previous PTransform we filter messages to only include docTypes of `topsites-impression` `topsites-click` `quicksuggest-impression` and `quicksuggest-click` and namespaces of `contextual-services` (desktop) and `org-mozilla-fenix` (phone). There will likely be additional namespaces (ios) but we're not currently supporting that and the form factor there will still be `phone`

I can move the ternary expressions to explicit conditionals if you feel like that would be a safer option.

Reply:
yeah, to explain by example, I'd expect something like:
```suggestion
          String[] docTypeParts = docType.split("-", 2);

          if (docTypeParts.length != 2) {
            throw new IllegalArgumentException("Invalid docType: " + docType);
          }

          String source = docTypeParts[0];
          String interactionType = docTypeParts[1];

          if (!SponsoredInteraction.SOURCE_TOPSITES.equals(source)
              && !SponsoredInteraction.SOURCE_SUGGEST.equals(source) {
            // Unclear if this is necessary
            throw new IllegalArgumentException("Invalid source: " + source);
          }
          // interactionType is checked in isUrlValid

          interactionBuilder.setSource(source);
          interactionBuilder.setInteractionType(interactionType);

          if (NS_DESKTOP.equals(namespace)) {
            interactionBuilder.setFormFactor(SponsoredInteraction.FORM_DESKTOP);
          } else {
            // All other namespaces are for mobile
            interactionBuilder.setFormFactor(SponsoredInteraction.FORM_PHONE);

            // Unclear if this is necessary
            ArrayNode events = payload.withArray("events");
            if (events.size() != 1) {
              throw new RuntimeException("expect exactly 1 event in ping.");
            }
            JsonNode event = events.get(0);
            String eventName = event.path("name").asText();
            if (!eventName.endsWith("-" + interactionType)) {
              throw new IllegalArgumentException("Invalid event name '" + eventName
                  + "' for interaction type '" + interactionType + "'");
            }
          }
```

Reply:
> In a previous PTransform we filter messages to only include docTypes of `topsites-impression` `topsites-click` `quicksuggest-impression` and `quicksuggest-click` and namespaces of `contextual-services` (desktop) and `org-mozilla-fenix` (phone). There will likely be additional namespaces (ios) but we're not currently supporting that and the form factor there will still be `phone`

If i understand correctly those filters are in configuration, rather than in code. If configuration changed this should handle that gracefully. For source and interaction type, we should use the content of docType split on `-`, rather than assume source and interaction will never be a substring of the other. It sounds like the desktop or phone choice is fine as is, because even if configuration changes, everything else is FORM_PHONE, so it's an accurate assumption.

Reply:
> I can move the ternary expressions to explicit conditionals if you feel like that would be a safer option.

It's not that the ternary is unsafe, it's more that I think the code should probably account for more than two outcomes. If new doc types are added later, then behavior like `if a {} else if b {} else {throw exception}` will prevent hard to find issues where a new option for source or interaction type gets hidden.

If checking the payload event name isn't needed then a ternary operator for form factor would make sense.

Reply:
> If new doc types are added later, then behavior like if a {} else if b {} else {throw exception} will prevent hard to find issues where a new option for source or interaction type gets hidden.

I agree with that and will address it.

> If checking the payload event name isn't needed then a ternary operator for form factor would make sense.
Unfortunately we only get one docType for mobile tiles (`topsites-impression`) and the interaction type is encoded in the event name. I'll add some comments around that so its clearer for future me/us.

Reply:
> i don't think contains is strict enough here

For example, if we later add a source like `siteclick` and a interaction type `click-impression`, then with `docType = "siteclick-click-impression"` we would get incorrect behavior from using `contains("click")` or `endsWith("-impression")`. If we want to use contains, then we have to enforce that none of these values are substrings of eachother.

So this could use `startsWith` and `endsWith` and prefix/suffix the match with `-` and enforce a single `-` in `docType`, or it could use a regex or `String.split` to ensure we don't do any partial matching.

> Unfortunately we only get one docType for mobile tiles (topsites-impression) and the interaction type is encoded in the event name. I'll add some comments around that so its clearer for future me/us.

we should also throw an exception when docType doesn't match that expectation

Reply:
I refactored this to use explicit `if {} else if {} else {}` and throwing an exception when the values don't match any expected value